[PrettyPrint] auto-pretty-print minified sources (#4171)

assafdd opened the pull request:
Associated Issue: #4171

### Summary of Changes
* copied isMinified heuristic from old debugger
* auto togglePrettyPrint on setText 

### Test Plan:
run tests on 
1. New Tab - about:home (some react  resources under vendor folder) 
2. http://devtools-html.github.io/debugger-examples/examples/todomvc/ (bower_components/backbone/local-storage folder)

One odd is that if you close  pretty-print source while it's selected it will reappear while if you close it while the original version is visible it won't.

assafdd commented:
the commit also broke Editor.spec.js with to the following exception
  ● Editor › should set text
    TypeError: Cannot read property 'match' of undefined
      at isOriginalId (node_modules/devtools-source-map/src/utils/index.js:19:14)

jasonLaster commented on src/components/Editor/index.js:
```diff
@@ -488,6 +492,15 @@ class Editor extends PureComponent<Props, State> {
+        isMinified(selected.id, selected.text)
+      ) {
+        togglePrettyPrint(selectedSource.get("id"));
+      }
```
can we do this in `actions/sources#selectSource` we like to have the actions control the business logic

jasonLaster commented on src/components/Editor/index.js:
```diff
@@ -224,7 +226,9 @@ class Editor extends PureComponent<Props, State> {
-    this.setText(nextProps);
+    if (nextProps.selectedSource !== this.props.selectedSource) {
+      this.setText(nextProps);
+    }
```
why do we need this?

bomsy replied:
It was resetting the text each time, we setEmptylines and other stuff even though the selected source has not changed

assafdd commented:
Move the login to the action.
@bomsy I don't think the if statement is needed any more. WDYT? 
`if (nextProps.selectedSource !== this.props.selectedSource) {`

assafdd commented:
Still some jest test are failing on my machine.

jasonLaster commented:
kill the IF

bomsy commented:
@assafdd yeah ..not relevant

assafdd commented:
@jasonLaster I removed the IF, fixed expressions.spec.js 
I can't figures out why createPrettySource.spec.js is breaking and if it's releted to my changes at all.
Any suggestions how to continue?

jasonLaster commented:
will look

bomsy commented on src/actions/sources.js:
```diff
@@ -207,6 +207,10 @@ export function selectSource(id: string, options: SelectSourceOptions = {}) {
         await dispatch(loadSourceText(source.toJS()));
         await dispatch(setOutOfScopeLocations());
+        const src = getSource(getState(), id).toJS();
+        if (src.text && isMinified(src.id, src.text)) {
```
We should also be checking if this is a generated file or not before auto pretty printing. That's probably the issue with the tests

bomsy replied:
adding  `&&  !isOriginalId(src.id)` should do help

assafdd replied:
thanks @bomsy I'm on it

assafdd replied:
done

bomsy commented:
tiny lint issues
![image](https://user-images.githubusercontent.com/792924/32904466-4ca9aa0c-caef-11e7-9b4f-80e8ca632b5b.png)

assafdd commented:
Well, I think it's finally ready :)

bomsy commented on src/actions/sources.js:
```diff
@@ -207,6 +208,10 @@ export function selectSource(id: string, options: SelectSourceOptions = {}) {
         await dispatch(loadSourceText(source.toJS()));
         await dispatch(setOutOfScopeLocations());
+        const src = getSource(getState(), id).toJS();
+        if (src.text && !isOriginalId(src.id) && isMinified(src.id, src.text)) {
```
hey @assafdd,
 just noticed one last tiny thing. we should do add `shouldPrettyPrint(src)`. it'll look something like this
```js
if (src.text && !isOriginalId(src.id) && (isMinified(src.id, src.text) || shouldPrettyPrint(src))) { 
```
 so we make sure all holes are covered.

jasonLaster replied:
could we do `shouldPrettyPrint(src)) && isMinified(src.id, src.text)`

bomsy replied:
That might be to much as there might be files that don't meet the `isMinified` heuristics but we might still want to pretty print if `shouldPrettyPrint(src)` is `true`

assafdd replied:
Yes we can. I move the empty text check to isMinified util. 
Otherwise "should evaluate expressions in specific scope" in expressions.spec.js breaks.

assafdd replied:
@bomsy I tried to play  with shouldPrettyPrint()  and its seems to be truthy on any JS source that I have checked.

bomsy replied:
cool! it's fine, i had a scenario which i found but can't find it any more so its fine

bomsy commented:
Thanks! @assafdd
